Initial implementation of the WebAssembly backend.

sjrd opened the pull request:
This commit contains the initial implementation of the WebAssembly backend. This backend is still experimental, in the sense that:

* We may remove it in a future Minor version, if we decide that it has a better place elsewhere, and
* Newer minor versions may produce WebAssembly code that requires more recent WebAssembly features.

The WebAssembly backend silently ignores `@JSExport` and `@JSExportAll` annotations. It is otherwise supposed to support the full Scala.js language semantics.

Currently, the backend only supports some configurations of the linker. It requires:

* No optimizer,
* Unchecked semantics for undefined behaviors,
* Strict floats, and
* ES modules.

Some of those will be relaxed in the future, definitely including the first two.

Co-authored-by: @tanishiking 

---

Past history of the development of this backend may be found at https://github.com/tanishiking/scala-wasm
We did not make any effort in preserving the git history when creating this PR, as it followed a very aggressive commit policy, which would not fit this repository anyway.

sjrd commented:
The build is now green. I have started some Readme documentation of the compilation scheme, to complement the code comments which are more local. I'll continue in the coming days.

@gzm0 I expect reviewing this will take time (like, weeks or more). Beside the compilation scheme, are there things you would like to see and that would help you review?

I'm already marking you as reviewer, but it's fine if you wait for me to provide everything you will need before starting, obviously.

gzm0 commented:
> Beside the compilation scheme, are there things you would like to see and that would help you review?

I think for now, this is OK. My plan was to first focus on the "outer structure" anyways, so for this, I do not need to know a lot (anything?) about the compilation scheme.

One thing that I'd like to get out of this PR is to get a better understanding how reviews will work going forward in this repository between @sjrd, @tanishiking and me. Basically, I'd like to understand if we need more "formal" structure to ensure that:
- Everybody stays in the loop about changes relevant to their area of expertise.
- We do not waste resources by having multiple people look at changes (I have a feeling this already happens sometimes between @sjrd and me when we have external contributions).

In any case though, I think we'll be better equipped at the end of the review of this PR to answer this question. I just wanted to put it out there, so we do not forget.

> from now on I will refrain from force-pushing on this branch

FWIW: For me Github's "viewed" state is likely going to be much more important than individual commits. So for example, feel free to amend the `README.md` comment as much as you'd like until it is ready for review.

gzm0 reviewed (changes requested):
First review pass on everything but `wasmemitter` and `webassembly`.

gzm0 commented:
One thing @sjrd you might be able to provide that helps me, is a "suggested order of review" for the two new packages. (I was assuming `webassembly` should go first, but maybe you have a different, more informed suggestion).

sjrd commented on test-suite/js/src/test/scala/org/scalajs/testsuite/library/LinkingInfoTest.scala:
```diff
@@ -16,16 +16,23 @@ import scala.scalajs.LinkingInfo
 class LinkingInfoTest {
-  @Test def productionMode(): Unit =
+  @Test def productionMode(): Unit = {
+    assumeFalse("Not implemented yet on WebAssembly", Platform.executingInWebAssembly)
```
Ah ah, yes, probably!

sjrd replied:
`set Global/enableWasmEverywhere := true` turns all the JS projects of the build into Wasm projects, including the examples. 😉

sjrd commented on linker-interface/shared/src/main/scala/org/scalajs/linker/interface/StandardConfig.scala:
```diff
@@ -177,6 +186,38 @@ final class StandardConfig private (
+   *
+   *  @note
+   *    The WebAssembly backend silently ignores `@JSExport` and `@JSExportAll`
+   *    annotations. All other language features are supported.
```
I think at link time we should keep ignoring it. Otherwise we would have to hard-fail for the exports of `jl.Object.toString()` and `Throwable.{name,message}`. More generally it might be fine for some use cases to ignore them.

That said, I was thinking we could add a compiler option that warns about them. This way a codebase that actively uses `@JSExport` would be warned against it. I'm not yet sure about the best thing to do.

gzm0 replied:
OK. Let's defer that discussion for later. But it's maybe worth adding a note that this behavior might change in the future?

sjrd commented on test-suite-ex/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/ClassTestEx.scala:
```diff
@@ -26,6 +29,8 @@ class ClassTestEx {
    *  pollute the whole test suite with those.
    */
   @Test def getSuperclass(): Unit = {
+    assumeFalse("Not supported on WebAssembly", executingInWebAssembly)
```
You're right. There's no blocker to support it.

sjrd commented on project/Build.scala:
```diff
@@ -2713,6 +2745,17 @@ object Build {
+      /* scala.util.RandomTest.testShuffle() on WebAssembly manages to crash V8
+       * with a SIGSEGV, which probably indicates a bug in V8 and not in our own
+       * backend.
+       */
```
In our issue tracker? Or in V8? For the latter I would need to figure out a minimal reproduction.

gzm0 replied:
> In our issue tracker?

Yes. I realize that filing this for V8 is potentially an entire project on its own.

Of course it does raise the question of how to track WASM work in our issue tracker, but we'll need that anyways (and it's probably as straight-forward as creating a WASM label).

sjrd replied:
https://github.com/scala-js/scala-js/issues/4990

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/DerivedIRFiles.scala:
```diff
@@ -0,0 +1,164 @@
+    derivedIRFiles.map(_.flatten)
+  }
+
+  /** Generates the accompanying Box class of `Character` or `Long`.
```
That is an interesting thought! I was sure we needed analyzer support for the default/reflective bridges to be correctly generated as well. But in fact they would also be created on the hijacked class, so we would also receive them for free.

I'll try. I sure would prefer to get rid of the `deriveInjectedIRFiles`!

gzm0 replied:
CI suggests its works :) Great!

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/WebAssemblyLinkerBackend.scala:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    def writeWasmFile(): Future[Unit] = {
+      val emitDebugInfo = !config.minify
```
IMO this is covered by the existing doc of `minify`, namely:

> When enabled, the linker more aggressively reduces the size of the generated code, at the cost of readability and debuggability. It does not perform size optimizations that would negatively impact run-time performance.

It reduces code size, does not affect run-time performance, but definitely worsens debuggability.

But I can make it more explicit.

tanishiking replied:
FYI: As you might find, `emitDebugInfo = false` result in not generating [custom 'name' section](https://webassembly.github.io/spec/core/appendix/custom.html#name-section) in Wasm, whose purpose is ...

> The purpose of this section is to attach printable names to definitions in a module, which e.g. can be used by a debugger or when parts of the module are to be rendered in [text form](https://webassembly.github.io/spec/core/text/index.html#text).

I believe this is the only minifyable thing in Wasm binary at this moment.

(Maybe it's reasonable to remove sourceMap when `minify = true` ?)

gzm0 replied:
Fair enough.

> But I can make it more explicit.

Probably not necessary. We probably do not want WASM / non-WASM specifics all over unless really necessary.

gzm0 replied:
> (Maybe it's reasonable to remove sourceMap when minify = true ?)

No, I don't think so:

- There is an explicit config option for this.
- It goes into different files, so for shipping / bundling it is a very different story (it is entirely reasonable/feasible to keep the source maps somewhere completely separate).

sjrd commented:
> One thing @sjrd you might be able to provide that helps me, is a "suggested order of review" for the two new packages. (I was assuming `webassembly` should go first, but maybe you have a different, more informed suggestion).

Good point. I would suggest the following:

In `webassembly`:

1. Start with the set `Identities`, `Types`, `Modules`, `Instructions`. They define our core immutable model of WebAssembly.
2. `ModuleBuilder`
3. `FunctionBuilder`, but ignore the body of `def switch` as well as `def localDeadCodeEliminationOfInstrs` (you can come back to those later after getting more comfortable with straightforward Wasm codegen)

This should be enough of `webassembly` to understand the rest. `BinaryWriter` and `TextWriter` can be reviewed independently of everything else. They can be good "lighter review" to do when you don't have much time to commit.

In `wasmemitter`:

1. Start with `EmbeddedConstants` and `SpecialNames` (they are trivial but some of their comments give context)
2. Review `TypeTransformer`
3. Read the doc comments of `VarGen.genFieldID.typeData` and its type members
4. Read `CoreWasmLib.genNewArrayOfThisClass` to get familiar with a) our "DSL" to emit Wasm instructions and b) what working with Wasm instructions looks like. I suggest this one because a) it is a direct equivalent of `newArrayOfThisClass` in `CoreJSLib`, b) it manipulates virtually no larger concept of our compilation scheme and c) it is non-trivial nevertheless.
5. Read `LoaderContent` once (or the *generated* `__loader.js` file for a project like the helloworld) to get an idea of what functions exist in there.
6. Read the headers of the other `def genX` methods of `CoreWasmLib` (not their implementation), to also get an idea of what helper functions exist.
7. At this point you should have enough overall context to get to get to the meat, so:
8. Read the "API" of `WasmContext` and `WasmContext.ClassInfo`.
9. Review `FunctionEmitter`. Keep `genApplyWithDispatch` for the end. Only do the "HERE BE DRAGONS" part if you feel adventurous.
10. Review `SWasmGen`
11. `ClassEmitter`, but ignore the monster `genCreateJSClassFunction`
12. `Emitter`
13. `CoreWasmLib`
14. `WasmContext`
15. `Preprocessor`
16. and finally what's left in no particular order

tanishiking commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Emitter.scala:
```diff
@@ -0,0 +1,403 @@
+    factory.multiple(
+      factory.instantiateClass(ClassClass, ClassCtor),
+
+      // TODO Ideally we should not require this, but rather adapt to its absence
+      factory.instantiateClass(JSExceptionClass, JSExceptionCtor),
```
I didn't expect `JSExceptionClass` to be in symbol requirements, why is this required? (what problem gonna be there without this?)

sjrd replied:
It turned necessary for the `helloworld` build to work. I think it only happens if the codebase contains a `try..xatch` but no `throw`.

tanishiking replied:
> try..catch but no throw.

ah, that makes sense, thanks!

tanishiking reviewed (commented):
I haven't looked at the code for the compilation scheme much, only checking the differences from the book's repository. (Since I've reviewed the most of code there through the interaction in scala-wasm repo).
I left a few comments but it looks good to me, and `wasmemitter/README.md` looks great so far.

gzm0 commented on test-bridge/src/main/scala/org/scalajs/testing/bridge/HTMLRunner.scala:
Regarding the changes in this file (I forgot on my first review):

- It feels like this is a bugfix that is valuable on its own. Maybe split it out and merge it already?
- It suggests that we can use `HTMLRunner` to test Wasm (which is cool). Add it to `TESTING.md`?

sjrd replied:
Both done.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```diff
@@ -0,0 +1,3242 @@
+    r.tpe
+  }
+
+  private def genThis(t: This): Type = {
+    markPosition(t)
+    genReadStorage(receiverStorage)
+    t.tpe
+  }
```
Yes, this was fixed upstream in the meantime: 
https://github.com/scala-js/scala-js/pull/4983

gzm0 commented:
FYI, I'll mark comments I made as resolved once I feel the discussion is resolved. @tanishiking @sjrd do you feel we need to agree on a scheme on how to resolve comments? (e.g. opener only resolves?)

sjrd commented:
I think opener resolves, unless it's a typo or something like that.

sjrd commented on Jenkinsfile:
```diff
@@ -393,6 +393,22 @@ def Tasks = [
         ++$scala $testSuite$v/test
   ''',
 
+  "test-suite-webassembly": '''
```
I added a commit that integrates more of what `main` tests, in particular the ability to run the helloworld and build the reversi/testingExample.

sjrd commented on javalib/src/main/scala/java/lang/StackTrace.scala:
```diff
@@ -64,6 +64,11 @@ private[lang] object StackTrace {
+      /* If the throwable is sealed, which notably happens on Wasm, we will not
+       * be able to capture the stack trace in it, so we also create a JS Error.
+       */
+      new js.Error()
```
Done and integrated into the original commit.

sjrd commented:
I believe I have addressed everything so far.

* The `createJSError` fix was integrated in its original commit, to keep being a stand-alone.
* I now squashed the change to deriving `LinkedClass`es into the big commit, since that actively reduces the diff.
* The rest is in small separate commits for now.

tanishiking reviewed (approved):
I know the README is WIP, but LGTM from my side :100:

sjrd commented:
I think I'm done with the README at this point. There are many more things I could write about, but they are more local, so comments in the code should cover them well. I focused on things that have a global impact across the backend.

gzm0 reviewed (changes requested):
Review of

- Modules.scala
- Identities.scala
- README.md

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/Types.scala:
```diff
@@ -0,0 +1,174 @@
+      id: TypeID,
+      originalName: OriginalName,
+      isFinal: Boolean,
+      superType: Option[TypeID],
```
What is this? I cannot find any reference to it in the spec, but the TextWriter / BinaryWriter seem to use it.

sjrd replied:
It's the `typeidx*` of the spec at https://webassembly.github.io/gc/core/syntax/types.html#recursive-types

There is an additional restriction that there is at most one, and it is interpreted as a super type so we gave it a better name here.

Note that the `id` also ultimately corresponds to a `typeidx`, but that one is implicitly allocated by the order of definitions in the types section.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/Instructions.scala:
```diff
@@ -0,0 +1,395 @@
+object Instructions {
+
+  /** A WebAssembly `expr`. */
+  sealed case class Expr(instr: List[Instr])
```
```suggestion
  final case class Expr(instr: List[Instr])
```

sjrd replied:
Done, as well as on all other `case class`es of this file.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/Instructions.scala:
```diff
@@ -0,0 +1,395 @@
+  // Fake instruction to mark position changes
+  final case class PositionMark(pos: Position) extends Instr("pos", -1)
+
+  // Unary operations
```
Would it make sense to order all instruction definitions by opcode so they are easily referenceable in https://webassembly.github.io/gc/core/appendix/index-instructions.html ?

sjrd replied:
Ah yes, that's probably a good idea. It grew somewhat organically, but the opcode order should be reasonably clean.

gzm0 reviewed (changes requested):
Review of `Types.scala` (and two stray comments on `Instructions.scala`, but I'm not done there, I realize it'll take me a while, so I wanted to release what I have already).

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/Modules.scala:
```diff
@@ -0,0 +1,119 @@
+      typeID: TypeID,
+      params: List[Local],
+      results: List[Type],
+      locals: List[Local],
```
Yes, that's right.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/Modules.scala:
```diff
@@ -0,0 +1,119 @@
+    final case class Global(id: GlobalID, originalName: OriginalName, tpe: Type, isMutable: Boolean)
+        extends ImportDesc
+
+    final case class Tag(id: TagID, originalName: OriginalName, typeID: TypeID) extends ImportDesc
```
It's used in `CoreWasmLib.genTags`. In fact the only tag that we use is an import for `WebAssembly.JSTag`.

gzm0 replied:
Right.. I seem to have missed that.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/Modules.scala:
```diff
@@ -0,0 +1,119 @@
+      val start: Option[FunctionID],
+      val elems: List[Element],
+      val datas: List[Data]
+  )
```
It's in the order of first appearance in the binary format. The "structure" format of the spec makes little sense to me, TBH, but we can also switch to that if you prefer.

gzm0 replied:
Ah, right. No, the binary format order makes more sense. I've added a suggestion to add this information to the comment.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/README.md:
```diff
@@ -0,0 +1,939 @@
+  (param $this (ref $c.A)) (result i32)
+  ;; field selection: push the object than `struct.get`
+  local.get $this
+  struct.get $c.A $f.A.x
```
Yes, indeed. `f` makes a lot of sense for fields, and for methods it matches the equivalent namespace in the JS backend. But if it's confusing we can change the latter.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/README.md:
```diff
@@ -0,0 +1,939 @@
+struct.get $c.java.lang.Object $itables
+i32.const 0 ;; Intf was assigned slot 0 in this example
+array.get $itables
+;; since multiple interfaces can be assigned to the same slot,
```
Ah I see the confusion. There's one indirection missing in the explanation. If we had only one interface per slot, we would not use an array; we would use a `struct` instead, where each slot has the precise type of its interface.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/README.md:
It's not a typo. `exn` is the contraction of `exception` in this context.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/Modules.scala:
```diff
@@ -0,0 +1,119 @@
+    }
+  }
+
+  /** A WebAssembly `module`. */
```
```suggestion
  /**
    * A WebAssembly `module`.
    *
    * Order of fields is in the order of the binary format:
    * https://webassembly.github.io/gc/core/binary/modules.html#sections
    */
```

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/Modules.scala:
```diff
@@ -0,0 +1,119 @@
+
+  object ExportDesc {
+    final case class Func(id: FunctionID, originalName: OriginalName) extends ExportDesc
+    final case class Global(id: GlobalID, originalName: OriginalName) extends ExportDesc
```
No, you're right. They shouldn't have an `originalName`.

---

In the current compilation scheme, we indeed don't use any `Export` at this point. However we did use them before and will likely use them again in the future, so I suggest we keep them in, in order not to lose their implementation in the builders and writers. I'm adding a comment about that.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/Modules.scala:
```diff
@@ -0,0 +1,119 @@
+    sealed abstract class Mode
+
+    object Mode {
+      case object Passive extends Mode
```
Indeed. I removed all of these small unused things, whether in code or in comments. (except `Export`, for the reason I added in a comment)

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/Modules.scala:
```diff
@@ -0,0 +1,119 @@
+    final case class Tag(id: TagID, originalName: OriginalName, typeID: TypeID) extends ImportDesc
+  }
+
+  /** A WebAssembly `func`, including names for parameters and locals. */
```
Applied.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/Modules.scala:
```diff
@@ -0,0 +1,119 @@
+      originalName: OriginalName,
+      tpe: Type,
+      init: Expr,
+      isMutable: Boolean
```
Indeed. I actually put it before `tpe`, which is what is truly consistent with the spec. I did the same in `ImportDesc.Global`.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/Identitities.scala:
```diff
@@ -0,0 +1,41 @@
+
+  trait DataID
+
+  trait ExportID
```
Removed.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/Identitities.scala:
```diff
@@ -0,0 +1,41 @@
+
+  trait TagID
+
+  trait DataID
```
Reordered.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/Types.scala:
```diff
@@ -0,0 +1,174 @@
+  /** A WebAssembly `storagetype`. */
+  sealed trait StorageType
+
+  /** A WebAssembly `valtype`. */
```
Comment added.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/Types.scala:
```diff
@@ -0,0 +1,174 @@
+    case object None extends AbsHeapType("none", "nullref", 0x71)
+    case object NoExtern extends AbsHeapType("noextern", "nullexternref", 0x72)
+    case object NoFunc extends AbsHeapType("nofunc", "nullfuncref", 0x73)
+    case object NoExn extends AbsHeapType("noexn", "nullexnref", 0x74)
```
Reordered to match the binary format.

sjrd commented:
I addressed the comments made on the code so far, as they might impact further review comments (notably in `Instructions`).

I'll address the comments on the Readme later this week.

gzm0 reviewed (changes requested):
Review of the full `webassembly` package excluding the two writers.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/Instructions.scala:
```diff
@@ -0,0 +1,403 @@
+  case object RefIsNull extends SimpleInstr("ref.is_null", 0xD1)
+  final case class RefFunc(i: FunctionID) extends FuncInstr("ref.func", 0xD2, i)
+  case object RefEq extends SimpleInstr("ref.eq", 0xD3)
+  case object RefAsNotNull extends SimpleInstr("ref.as_non_null", 0xD4)
```
Ouch 🙈 Nice catch.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/FunctionBuilder.scala:
```diff
@@ -0,0 +1,406 @@
+    require(clauseSig.params.size >= scrutineeSig.results.size,
+        "The clauses of a switch must consume all the results of the scrutinee " +
+        s"(scrutinee results: ${scrutineeSig.results}; clause params: ${clauseSig.params})")
+    val (doneBlockType, clauseBlockType) = {
```
Hum, could you elaborate? I don't understand what you mean in this comment.

gzm0 replied:
Ah, I meant something like this (for the whole function):

```scala
val numCases = clauses.map(_._1.max).max + 1
require(numCases < 128, "...")

val doneLabel = genLabel()
val defaultLabel = genLabel()
val clauseLabels = clauses.map(_ => genLabel())

val dispatchVector = {
  val dv = Array.fill(numCases)(defaultLabel)
  for {
    (clauses, clauseLabel) <- clauses.zip(clauseLabels)
    (caseValue, _) <- clauses
  } {
    require(dv(caseValue) == defaultLabel), "...")
    dv(caseValue) = clauseLabel
  }
  dv.toList
}

// input parameter to the overall switch "instruction"
val switchInputParams =
  clauseSig.params.drop(scrutineeSig.results.size) ::: scrutineeSig.params

val doneBlockType = sigToBlockType(FunctionType(switchInputParams, clauseSig.results))
val clauseBlockType = sigToBlockType(FunctionType(switchInputParams, clauseSig.params))

// Open done block
instrs += Block(doneBlockType, Some(doneLabel))
// Open case and default blocks (in reverse order: default block is outermost!)
for (label <- (defaultLabel :: clauseLabels.reverse)) {
  instrs += Block(clauseBlockType, Some(label))
}

// Load the scrutinee and dispatch
scrutinee()
instrs += BrTable(dispatchVector, defaultLabel)

// Close all the case blocks and emit their respective bodies
for ((_, caseBody) <- clauses) {
  instrs += End // close the block whose label is the corresponding label for this clause
  caseBody() // emit the body of that clause
  instrs += Br(doneLabel) // jump to done
}

// Close the default block (no jump to done necessary)
instrs += End
default()

instrs += End // close the done block
```

sjrd replied:
Ah thanks, that clears it up. I integrated your version.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/FunctionBuilder.scala:
```diff
@@ -0,0 +1,406 @@
+        while (nestingLevel >= 0 && iter.hasNext) {
+          val deadCodeInstr = iter.next()
+          deadCodeInstr match {
+            case End | Else | _: Catch if nestingLevel == 0 =>
```
Ah yes, indeed, the spec for the legacy mechanism is hard to find. It's here: https://webassembly.github.io/exception-handling/legacy/exceptions/core/syntax.html

We will indeed get rid of it as soon as we get a version of Node.js that correctly handles `JSTag` in the new exception system.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/FunctionBuilder.scala:
```diff
@@ -0,0 +1,406 @@
+      BlockType.FunctionType(moduleBuilder.functionTypeToTypeID(sig))
+  }
+
+  def ifThenElse(blockType: BlockType)(thenp: => Unit)(elsep: => Unit): Unit = {
```
The latest commit uses a typeclass instead. There's still an overload for the empty block type, and `switch` still needs its overloads because it wants `FunctionType`s. Is it better?

sjrd commented:
Latest comments addressed.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/FunctionBuilder.scala:
```diff
@@ -0,0 +1,406 @@
+   *  `StructuredLabeledInstr`s during that process, must jump over them. That means we need to
+   *  track the level of nesting at which we are.
+   */
+  private def localDeadCodeEliminationOfInstrs(): List[Instr] = {
```
I considered that, but every way I looked at it, it ended up being much worse. The insert mechanism is one issue. What's even more problematic is maintaining state for the sub-blocks in eliminated code. At least this way, all the logic stays within this one method.

tanishiking commented:
@xuwei-k found that Wasm backend crashes with `Enumeration` 🤔 

```scala
$ cat examples/helloworld/src/main/scala/helloworld/HelloWorld.scala
/* Scala.js example code
 * Public domain
 * @author  Sébastien Doeraene
 */

package helloworld

object HelloWorld {
  def main(a: Array[String]): Unit = {
    println(Foo.A1)
  }
}

object Foo extends Enumeration {
  val A1 = Value
}
```

```sh
sbt:Scala.js> set Global/enableWasmEverywhere := true
sbt:Scala.js> helloworld2_13/fastLinkJS
[info] Fast optimizing /Users/tanishiking/ghq/github.com/tanishiking/scala-js/examples/helloworld/.2.13/target/scala-2.13/helloworld-fastopt
[error] stack trace is suppressed; run last helloworld2_13 / Compile / fastLinkJS for the full output
[error] (helloworld2_13 / Compile / fastLinkJS) java.util.concurrent.ExecutionException: Boxed Error
[error] Total time: 1 s, completed Jun 9, 2024, 9:42:49 AM

[error] Caused by: java.lang.Error: Class not found: ClassName<scala.Enumeration$ValueSet>
[error]         at org.scalajs.linker.backend.wasmemitter.WasmContext.$anonfun$getClassInfo$1(WasmContext.scala:80)
[error]         at scala.collection.immutable.HashMap$HashTrieMap.getOrElse0(HashMap.scala:596)
[error]         at scala.collection.immutable.HashMap$HashTrieMap.getOrElse0(HashMap.scala:589)
[error]         at scala.collection.immutable.HashMap.getOrElse(HashMap.scala:73)
[error]         at org.scalajs.linker.backend.wasmemitter.WasmContext.getClassInfo(WasmContext.scala:80)
[error]         at org.scalajs.linker.backend.wasmemitter.TypeTransformer$.transformClassType(TypeTransformer.scala:83)
[error]         at org.scalajs.linker.backend.wasmemitter.TypeTransformer$.transformType(TypeTransformer.scala:70)
[error]         at org.scalajs.linker.backend.wasmemitter.ClassEmitter.$anonfun$genScalaClass$3(ClassEmitter.scala:324)
[error]         at scala.collection.immutable.List.map(List.scala:297)
[error]         at org.scalajs.linker.backend.wasmemitter.ClassEmitter.genScalaClass(ClassEmitter.scala:320)
[error]         at org.scalajs.linker.backend.wasmemitter.ClassEmitter.genClassDef(ClassEmitter.scala:74)
[error]         at org.scalajs.linker.backend.wasmemitter.Emitter.$anonfun$emit$3(Emitter.scala:84)
[error]         at org.scalajs.linker.backend.wasmemitter.Emitter.$anonfun$emit$3$adapted(Emitter.scala:83)
[error]         at scala.collection.immutable.List.foreach(List.scala:431)
[error]         at org.scalajs.linker.backend.wasmemitter.Emitter.emit(Emitter.scala:83)
[error]         at org.scalajs.linker.backend.WebAssemblyLinkerBackend.emit(WebAssemblyLinkerBackend.scala:78)
[error]         at org.scalajs.linker.standard.StandardLinkerImpl.$anonfun$link$2(StandardLinkerImpl.scala:47)
[error]         at scala.concurrent.Future.$anonfun$flatMap$1(Future.scala:307)
[error]         at scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:41)
[error]         at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
[error]         at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
[error]         at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
[error]         at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
[error]         at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
[error]         at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
[error]         at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
[error] (helloworld2_13 / Compile / fastLinkJS) java.util.concurrent.ExecutionException: Boxed Error
```

sjrd commented:
Does cherry-picking https://github.com/sjrd/scala-js/commit/60a370112e2f79fa1daf339b41ec5148413ba395 solve the issue?

tanishiking commented:
> Does cherry-picking https://github.com/sjrd/scala-js/commit/60a370112e2f79fa1daf339b41ec5148413ba395 solve the issue?
@sjrd Oh yeah, it does solve the issue :+1:

gzm0 commented:
I've started reviewing `wasmemitter`. Just as a reminder, I believe the comments on the README are not addressed yet.

sjrd commented:
> Just as a reminder, I believe the comments on the README are not addressed yet.

Yes, indeed. All my week was focused on being ready to present at the Wasm CG meeting. 😅

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/README.md:
```diff
@@ -0,0 +1,939 @@
+
+This strategy trades off efficiency for space.
+It is slow, but that corresponds to the fact that reflective calls are slow on the JVM as well.
+In order to have fixed slots for reflective proxy methods, we would need a `O(m*n)` matrix where `m` is the number of concrete classes and `n` the number of distinct reflective proxy names in the entire program.
```
Ah ah, I think the current lookup is `O(n)`. We really should make it `O(log n)`, though. That's easy.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```suggestion
            // a primitive array always has the correct type
```

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
Nit: Put on previous line?

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
What are these implicits used for?

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
I do not understand why this instruction is necessary. Doesn't the `RefCast` already disallow `null`?

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
For reference: This is how I'd expect most `genX` methods to start.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```suggestion
          (matchableLiterals, label) <- caseLabels
```

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```suggestion
        for {
          (caseLabel, (_, caseBody)) <- caseLabels.zip(cases).reverse)
        } {
```

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
Nit: Call the helper `jsSuperSelect` for consistency?

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
TODO: Continue here.

gzm0 reviewed (changes requested):
Full review of function emitter up to dragons + review of review changes where I have follow-up.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/FunctionBuilder.scala:
```diff
@@ -0,0 +1,404 @@
+   */
+  BlockTypeLike.ForBlockType
+
+  def ifThenElse[BT: BlockTypeLike](blockType: BT)(thenp: => Unit)(elsep: => Unit): Unit = {
```
Ah ah, yes, it does compile. Although I'm not sure how I feel about the language letting that through. 😅

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```diff
@@ -0,0 +1,3241 @@
+      case t: TryFinally          => unwinding.genTryFinally(t, expectedType)
+      case t: Throw               => genThrow(t)
+      case t: Match               => genMatch(t, expectedType)
+      case t: Debugger            => NoType // ignore
```
I don't think so. Already in JS `debugger` is not guaranteed to do anything. I don't think it's incorrect to ignore it for Wasm altogether.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```diff
@@ -0,0 +1,3241 @@
+    val funcTypeID = ctx.tableFunctionType(t.method.name)
+
+    fb.block(watpe.RefType.anyref) { done =>
+      fb.block(watpe.RefType.any) { labelNotOurObject =>
```
Oh, I had forgotten that we had this TODO here. The fun fact is that it actually already works. We do have (plenty of) tests for reflective calls on primitives in `ReflectiveCallTest`. The thing is that the *compiler* already inserts dispatch for primitives when required. For example, the method
```scala
def foo(x: { def compareTo(that: Integer): Int }): Unit =
  println(x.compareTo(6))
```
has the following IR:
```scala
  def foo;Ljava.lang.Object;V(x: any) {
    mod:scala.Predef$.println;Ljava.lang.Object;V({
      val x$2: any = x;
      if (x$2.isInstanceOf[java.lang.Integer]) {
        x$2.asInstanceOf[java.lang.Integer].compareTo;Ljava.lang.Integer;I(mod:scala.Predef$.int2Integer;I;Ljava.lang.Integer(6).asInstanceOf[java.lang.Integer])
      } else {
        x$2.compareTo;Ljava.lang.Integer;R(mod:scala.Predef$.int2Integer;I;Ljava.lang.Integer(6).asInstanceOf[java.lang.Integer])
      }
    })
  }
```

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```diff
@@ -0,0 +1,3241 @@
+   */
+  private def genApplyWithDispatch(t: Apply,
+      receiverClassInfo: WasmContext.ClassInfo): Type = {
+    implicit val pos: Position = t.pos
```
The calls to `genUnbox`.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```diff
@@ -0,0 +1,3241 @@
+      val resultType = transformResultType(t.tpe)
+
+      fb.block(resultType) { labelDone =>
+        def pushArgs(argsLocals: List[wanme.LocalID]): Unit =
```
That doesn't work. We use both inside `val argsLocal = fb.block(...)` and *after* the `val argsLocal`. The argument is always textually `argsLocal` but they refer to two different `val`s. (those two `val`s end up always having the same value, but in different scopes).

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```diff
@@ -0,0 +1,3241 @@
+
+        BoxedClassToPrimType.get(targetClassName) match {
+          case None =>
+            genTree(t.receiver, ClassType(targetClassName))
```
In general I've chosen to leave NPE traps attached to the receiver, in terms of position handling. The rationale is to avoid too much back-and-forth when stepping through the code. Even though the NPE belongs to the application, in practice, having it report on the position of the receiver is just as good.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```diff
@@ -0,0 +1,3241 @@
+    fb += wa.Call(genFunctionID.arrayTypeData)
+  }
+
+  private def genClassOfFromTypeData(loadTypeDataInstr: wa.Instr): Unit = {
```
This whole thing was a mess, indeed. Always calling the `getClassOf` helper is probably better, indeed.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```diff
@@ -0,0 +1,3241 @@
+      }
+
+      tpe
+    }
```
I unnested the helpers but I kept the `binary` argument. It provides the positions and expected types (notably the position of the `rhs` when we only pass the `rhsValue`). Removing it requires passing even more information that looks duplicated, IMO.

I did add assertions on the `binary.op`s.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```diff
@@ -0,0 +1,3241 @@
+    }
+  }
+
+  private def genEq(binary: BinaryOp): Type = {
```
This one does get a lot more involved once we address the TODO about optimizing it. I have a commit about that in #4993.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```diff
@@ -0,0 +1,3241 @@
+          fb += wa.Call(genFunctionID.isString)
+
+        case testType: PrimTypeWithRef =>
+          testType match {
```
To be able to do `testType.primRef` in the `case _ =>`

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```diff
@@ -0,0 +1,3241 @@
+   *
+   *  The type left on the stack is non-nullable.
+   */
+  private def genUnbox(targetTpe: PrimType)(implicit pos: Position): Unit = {
```
It was used for `zeroOf` but I'm now using `genZeroOf`, which does not require the `pos`.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```diff
@@ -0,0 +1,3241 @@
+        fb += wa.RefAsNonNull
+
+      case targetTpe: PrimTypeWithRef =>
+        targetTpe match {
```
Same reply: for `targetTpe.primRef`.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```diff
@@ -0,0 +1,3241 @@
+    t.cond match {
+      case BooleanLiteral(true) =>
+        // infinite loop that must be typed as `nothing`, i.e., unreachable
+        markPosition(t)
```
I don't think this helps in this case, because of the additional `markPosition` in the middle, which cannot be factored out. Moving this one up makes it less obvious that everything is well-positioned, IMO.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```diff
@@ -0,0 +1,3241 @@
+            val resultType = transformType(targetTpe)
+
+            fb.block(Sig(List(watpe.RefType.anyref), List(resultType))) { doneLabel =>
+              fb.block(Sig(List(watpe.RefType.anyref), Nil)) { isNullLabel =>
```
It looks like it, but not really. There only 2 spots where the exact shape of the "then" branch performs an unconditional `Br(doneLabel)`. The other double-blocks that look like this one have partial fall-through of the then branch into the else branch. So I don't think this abstracts well.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```diff
@@ -0,0 +1,3241 @@
+          ),
+          List(StringLiteral("/ by zero"))
+        )
+      )
```
This one I addressed.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```diff
@@ -0,0 +1,3241 @@
+            )(
+              t.pos
+            )
+          )
```
I generally agree, but in this case this is really a lot simpler this way. One way to look at it is that we are forcing an optimization here. For an optimization, it makes sense to replace some IR by some other.

gzm0 replied:
Ah, fair enough. I presume we're also planning on removing this once we can enable the optimizer? If yes, maybe this is worth a comment?

sjrd commented:
I believe I have addressed everything so far.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```diff
@@ -0,0 +1,3349 @@
+)(implicit ctx: WasmContext) {
+  import FunctionEmitter._
+
+  private var innerFuncIdx: Int = 0
```
```suggestion
  private var closureIdx: Int = 0
```

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```diff
@@ -0,0 +1,3241 @@
+          watpe.RefType(genTypeID.ObjectStruct)
+        )
+        fb += wa.StructGet(
+          genTypeID.forClass(ObjectClass),
```
Actually, it looks like there are 3 other occurrences of this. Maybe it makes sense to remove `genTypeID.ObjectStruct` instead?

sjrd commented:
Rebased.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/TextWriter.scala:
```diff
@@ -0,0 +1,611 @@
+  private def writeImport(i: Import)(implicit b: WatBuilder): Unit = {
+    b.newLineList(
+      "import", {
+        b.appendElement("\"" + i.module + "\"")
+        b.appendElement("\"" + i.name + "\"")
```
Needs escaping. `module` and `name` are allowed to be anything.

https://webassembly.github.io/gc/core/text/values.html#strings

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/TextWriter.scala:
```diff
@@ -0,0 +1,611 @@
+      case FunctionType(params, results) =>
+        b.sameLineList(
+          "func", {
+            params.foreach { ty =>
+              b.sameLineList("param", writeType(ty))
+            }
+            results.foreach { ty =>
+              b.sameLineList("result", writeType(ty))
+            }
```
```suggestion
            writeSig(params, results)
```

Alternative, remove `writeSig` (which seems unused).

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/TextWriter.scala:
```diff
@@ -0,0 +1,611 @@
+    b.newLineList(
+      "func", {
+        appendName(f.id)
+        writeTypeUse(f.typeID)
```
Nit: IIUC, if we need to distinguish definition and usage, we use `Def` as a suffix for definition and no suffix for usage.

Or maybe here, `writeTypeID` is an appropriate name?

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/TextWriter.scala:
```diff
@@ -0,0 +1,611 @@
+
+        b.newLine()
+        f.params.foreach(writeParam)
+        f.results.foreach(r => { b.sameLineList("result", writeType(r)) })
```
```suggestion
        f.results.foreach(r => b.sameLineList("result", writeType(r)))
```

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/TextWriter.scala:
```diff
@@ -0,0 +1,611 @@
+    )
+  }
+
+  private def writeGlobal(g: Global)(implicit b: WatBuilder) =
+    b.newLineList(
+      "global", {
+        appendName(g.id)
+        if (g.isMutable)
+          b.sameLineList("mut", writeType(g.tpe))
+        else writeType(g.tpe)
+        g.init.instr.foreach(writeInstr)
+      }
+    )
```
```suggestion
  private def writeGlobal(g: Global)(implicit b: WatBuilder) = {
    b.newLineList(
      "global", {
        appendName(g.id)
        if (g.isMutable)
          b.sameLineList("mut", writeType(g.tpe))
        else writeType(g.tpe)
        g.init.instr.foreach(writeInstr)
      }
    )
  }
```

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/TextWriter.scala:
```diff
@@ -0,0 +1,611 @@
+      }
+    )
+
+  private def writeExport(e: Export)(implicit
+      b: WatBuilder
+  ) = b.newLineList(
+    "export", {
```
```suggestion
  private def writeExport(e: Export)(implicit b: WatBuilder) = {
    b.newLineList(
      "export", {
```

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/TextWriter.scala:
```diff
@@ -0,0 +1,611 @@
+      b: WatBuilder
+  ) = b.newLineList(
+    "export", {
+      b.appendElement("\"" + e.name + "\"")
```
Needs escaping.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/TextWriter.scala:
```diff
@@ -0,0 +1,611 @@
+        element.init.foreach { item =>
+          b.newLineList(
+            "item",
+            item.instr.foreach(writeInstr(_))
```
Add `writeExpr` as a method? (would also help `writeGlobal`).

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/TextWriter.scala:
```diff
@@ -0,0 +1,611 @@
+    }
+  }
+
+  private def floatString(v: Double): String = {
```
`writeFloatString`? This is only ever used to actually write the string. Returning it feels very inconsistent.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/TextWriter.scala:
```diff
@@ -0,0 +1,611 @@
+    blockType match {
+      case BlockType.FunctionType(typeID) =>
+        writeTypeUse(typeID)
+      case BlockType.ValueType(optType) =>
+        for (tpe <- optType)
+          b.sameLineList("result", writeType(tpe))
```
```suggestion
      case BlockType.ValueType(None) =>
        // do nothing
      case BlockType.ValueType(Some(tpe)) =>
        b.sameLineList("result", writeType(tpe))
```

IMO easier to read.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/TextWriter.scala:
```diff
@@ -0,0 +1,611 @@
+    }
+  }
+
+  private def writeLabelIdx(labelIdx: LabelID)(implicit b: WatBuilder): Unit =
+    appendName(labelIdx)
```
Any reason this is different from `appendName`?

sjrd replied:
Ah indeed. No I think this was historical.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/TextWriter.scala:
```diff
@@ -0,0 +1,611 @@
+    instr match {
+      case PositionMark(_) =>
+        // ignore
+        ()
```
I'm not sure that in general in the SJS codebase we write out units explicitly like this.

But you are the code style arbiter :)

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/TextWriter.scala:
```diff
@@ -0,0 +1,611 @@
+
+      case I32Const(v) => b.appendElement(v.toString())
+      case I64Const(v) => b.appendElement(v.toString())
+      case F32Const(v) => b.appendElement(floatString(v.toDouble))
```
Add TODO for strict floats?

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/TextWriter.scala:
```diff
@@ -0,0 +1,611 @@
+        writeType(from)
+        writeType(to)
+
+      case PositionMark(pos) =>
```
```suggestion
      case PositionMark(_) =>
```

gzm0 reviewed (changes requested):
Next round of reviews. I have looked at all the files that have comments fully (except for the dragons in `FunctionEmitter`).

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/CoreWasmLib.scala:
```diff
@@ -0,0 +1,2206 @@
+          fb += StructGet(genTypeID.typeData, genFieldID.typeData.kind)
+        }(
+          List(KindBoolean) -> { () =>
+            fb += ctx.refFuncWithDeclaration(genFunctionID.clone(BooleanRef))
```
```suggestion
            fb += ctx.refFuncWithDeclaration(genFunctionID.cloneArray(BooleanRef))
```

It took me quite a while to figure out what was happening here.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/CoreWasmLib.scala:
```diff
@@ -0,0 +1,2206 @@
+        fb += RefNull(HeapType.NoFunc)
+
+        // reflectiveProxies
+        fb += ArrayNewFixed(genTypeID.reflectiveProxies, 0) // TODO
```
Why the TODO? IIUC there cannot be a reflective proxy on an array? (It only has methods from Object, which never need a reflective proxy).

tanishiking replied:
Oh, right, I forgot to remove it :bow:

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/SWasmGen.scala:
```diff
@@ -0,0 +1,107 @@
+object SWasmGen {
+
+  def genBooleanNot(fb: FunctionBuilder): Unit = {
+    fb += I32Const(1)
+    fb += I32Xor
```
Can't we use `I32Eqz` for this? IIUC the CoreWasmLib does that already. (means its probably not worth having the helper?)

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/CoreWasmLib.scala:
```diff
@@ -0,0 +1,2206 @@
+      fb.block(RefType.any) { nonNullLabel =>
+        fb += LocalGet(valueParam)
+        fb += BrOnNonNull(nonNullLabel)
+        fb += GlobalGet(genGlobalID.bFalse)
```
Argh, this confused me, because for normal paths, we box at the end. Not clear if there is something we can do about it though.

Probably OK to leave it as is.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/CoreWasmLib.scala:
```diff
@@ -0,0 +1,2206 @@
+     * Leave it on the stack.
+     */
+    fb += Call(genFunctionID.jsNewObject)
+    // "__typeData": typeData (TODO hide this better? although nobody will notice anyway)
```
Add a comment why we need this? (IIUC for `isAssignableFromExternal`).

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/CoreWasmLib.scala:
```diff
@@ -0,0 +1,2206 @@
+   *
+   *  Specified by `java.lang.Class.isAssignableFrom(Class)`.
+   */
+  private def genIsAssignableFrom()(implicit ctx: WasmContext): Unit = {
```
Note to self: Continue here.

gzm0 reviewed (changes requested):
Review of Emitter, SWasmGen and some of CoreWasmLib.

I haven't gotten to WasmContext yet, I'll point out how I'd suggest to split it when I get there.

One thing I found challenging to understand is the order in which things need to be emitted. I realize I do not understand what requirements WASM has in that regard. Do you think this could be something worth adding to the README?

gzm0 reviewed (changes requested):
Complete review of CoreWasmLib

tanishiking commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/ClassEmitter.scala:
```diff
@@ -0,0 +1,1247 @@
+
+  /** Generates code for a top-level export.
+   *
+   *  The strategy for top-level exports is as follows:
```
I remember this is for handling (top-level exported) mutable vars: https://github.com/tanishiking/scala-wasm/commit/e170e2bcf8a78bb514124299f53d779ba40f24f9

For example, when we have the following Scala code that exports mutable `bar`,

```scala
object Foo {
  @JSExportTopLevel("foobar")
  var bar = 42
}
```

The generated JS file looks like this:

```js
import { load as __load } from './__loader.js';

let exportedfoobar;
export { exportedfoobar as foobar };

await __load('./main.wasm', {

}, {
  foobar: (x) => exportedfoobar = x,
});
```

This `exportedfoobar` mirrors the state, and the source of truth for the state remains in the Wasm Global for the `Foo.bar`.

When the `Foo.bar` is updated in Wasm (by `Assign(SelectStatic(...), ...)`), we pushes the change to the mirror using the imported setter function `foobar: (x) => exportedfoobar = x`.

https://github.com/scala-js/scala-js/blob/b418fe165c91d220dad832fe7d292d604c009954/linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala#L502-L515

---

It'd be great to generate those mirrors only for mutable exports, but sure it's feasible 🤔

tanishiking commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Preprocessor.scala:
```diff
@@ -0,0 +1,391 @@
+   *    "Efficient Type Inclusion Tests"
+   *    [[https://www.researchgate.net/publication/2438441_Efficient_Type_Inclusion_Tests]]
+   */
+  private def assignBuckets(allClasses: List[LinkedClass],
```
I will add more detailed explanation to the doc comment + try to a bit of refactoring :+1:

gzm0 replied:
As a suggestion: There is nothing wrong with using an algorithm that uses highly mutable datastructures (especially if it is described in the literature as such, much easier to cross-check). However, unless it is absolutely necessary for performance (measure!), I'd try to keep the interfaces immutable and functionally pure. That makes it easy to reason about how the algorithm fits into the larger surroundings.

tanishiking commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/CoreWasmLib.scala:
```diff
@@ -0,0 +1,2206 @@
+  }
+
+  /** `createStringFromData: (ref array u16) -> (ref any)` (representing a `string`). */
+  private def genCreateStringFromData()(implicit ctx: WasmContext): Unit = {
```
> But why not just emit all the string constants in JS in an array and get them from there?

Because `array` in Wasm is opaque to JS 😞 only `i32`, `i64`, `f32`, `f64` can be exported to JS, and hence iterating over Wasm array and convert to JS char one by one, and concat them is the only way :P

A good news is that, there's a proposal called [js-string-builtins](https://github.com/WebAssembly/js-string-builtins) and it contains `"wasm:js-string" "fromCharCodeArray"` that convert `i16 array` to JS string.
https://github.com/WebAssembly/js-string-builtins/blob/main/proposals/js-string-builtins/Overview.md#wasmjs-string-fromcharcodearray
This should makes it much easier and faster to create JS string from Wasm array.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/ClassEmitter.scala:
```diff
@@ -0,0 +1,1247 @@
+        // Generate boxed representation of the zero of the field
+        fb += genBoxedZeroOf(fieldDef.ftpe)
+
+        fb += wa.Call(genFunctionID.installJSField)
```
That would add an additional helper for no reason, since it would do exactly the same thing as `installJSField`. I added a comment about that.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/ClassEmitter.scala:
```diff
@@ -0,0 +1,1247 @@
+      fb += wa.GlobalGet(cachedJSClassGlobal.id)
+      fb += wa.BrOnNonNull(doneLabel)
+      // Otherwise, call createJSClass -- it will also store the class in the cache
+      fb += wa.Call(genFunctionID.createJSClassOf(className))
```
See my other comment: we cannot do that.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/ClassEmitter.scala:
```diff
@@ -0,0 +1,1247 @@
+      fb += wa.BrOnNonNull(doneLabel)
+
+      // Get the JS class and instantiate it
+      fb += wa.Call(genFunctionID.loadJSClass(className))
```
TBH here I copied the strategy of the JS backend. I am not sure whether it is *necessary* to do it like this. That said, it is IMO the easiest to implement and the clearest.

sjrd reviewed (commented):
Addressed comments in `ClassEmitter` and added details about the overall ordering of definitions in the Readme.

Sorry for the delay in addressing the latest batch. I'm mostly on vacation these days.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Emitter.scala:
```diff
@@ -0,0 +1,403 @@
+
+    val loaderContent = LoaderContent.bytesContent
+    val jsFileContent =
+      buildJSFileContent(module, module.id.id + ".wasm", allImportedModules)
```
Why? The `.wasm` file name is already dictated almost by spec by the `module.id`, isn't it?

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Emitter.scala:
```diff
@@ -0,0 +1,403 @@
+      moduleInitializers: List[ModuleInitializer.Initializer],
+      topLevelExportDefs: List[LinkedTopLevelExport]
+  )(implicit ctx: WasmContext): Unit = {
+    /* Before generating the string pool in `genStringPoolData()`, make sure
+     * to allocate the ones that will be required by the module initializers.
+     */
+    for (init <- moduleInitializers) {
+      ModuleInitializerImpl.fromInitializer(init) match {
+        case ModuleInitializerImpl.MainMethodWithArgs(_, _, args) =>
+          args.foreach(ctx.addConstantStringGlobal(_))
+        case ModuleInitializerImpl.VoidMainMethod(_, _) =>
+          () // nothing to do
+      }
+    }
```
After doing all the other suggested rearrangements, I think it's not too bad to keep this one there.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Emitter.scala:
```diff
@@ -0,0 +1,403 @@
+       */
+      val exprs = funcDeclarations.map { funcID =>
+        wa.Expr(List(wa.RefFunc(funcID)))
+      }
```
Declarative elements are not true expressions. It's really a hack to declare `ref.func`s ahead of time. They pretty much have to follow that shape. There are shortcuts for a long list of func refs, but it's at the binary format level, not at the structural level.

gzm0 replied:
Note to self: I do not understand yet, why ahead of time declarations like this are even necessary.

sjrd replied:
Meh, they just are. 🤷‍♂️ All `ref.func`s appearing in the code section must have a "forward declaration" somewhere in the earlier sections. In the absence of an existing occurrence, a `declarative` `elem` is the canonical place to put it. 

Some context I found: https://github.com/WebAssembly/gc/issues/499
Notable quote:
> This forward declaration thing probably is one of Wasm's biggest hacks.

😆

I'm not sure what's the rationale behind this requirement, though.

sjrd reviewed (commented):
Addressed comments in `CoreWasmLib`, `DerivedClasses`, and some comments in `Emitter` but definitely not all of them yet.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/ClassEmitter.scala:
```diff
@@ -0,0 +1,1247 @@
+
+          case _ =>
+            // For everything else, put the tree in its own function and call it
+            val closureFuncID = new JSClassClosureFunctionID(className)
```
TY for the clarification. All-in-all I feel the name situation needs improvement, but I would not do so in this iteration / PR. I feel there are a couple of things that are tricky:

- We need debug names and wat names, which are subtly different (uniqueness, allowed charset)
- We want to construct "artificial" identifiers to identify the higher level construct that lead to the underlying WASM name. This means we actually have 3 input names for every element:
  - It's identifier.
  - It's generated artificial identifier
  - It's original name

In any case, I think we should leave it as-is in this PR to not explode the scope and reconsider / discuss later.

I'll resolve all other comments relating to this. @sjrd please resolve this one if you agree to leave as is and assess later.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/TextWriter.scala:
```diff
@@ -0,0 +1,611 @@
+}
+
+object TextWriter {
+  private class FreshNameGenerator {
+    private val generated = mutable.HashSet.empty[String]
+
+    def genName(originalName: OriginalName): String = {
+      val base =
+        if (originalName.isDefined) "$" + sanitizeWatIdentifier(originalName.get.toString())
+        else "$"
+      if (originalName.isDefined && generated.add(base)) {
+        base
+      } else {
+        var index = 1
+        while (!generated.add(base + index))
+          index += 1
+        base + index
+      }
+    }
+  }
```
Discussed in https://github.com/scala-js/scala-js/pull/4988#discussion_r1667337484

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/CoreWasmLib.scala:
```diff
@@ -0,0 +1,2206 @@
+    )
+    val itablesField = StructField(
+      genFieldID.objStruct.itables,
+      OriginalName(genFieldID.objStruct.itables.toString()),
```
Further discussed in https://github.com/scala-js/scala-js/pull/4988#discussion_r1667337484

tl;dr: I suggest we defer the discussion of what names we have and how they relate. For now, I think it stands to reason to *not* include names in IDs.

gzm0 reviewed (changes requested):
Review of most recent changes and answers to some comments.

I have also done a full "resolution" pass.

> Sorry for the delay in addressing the latest batch. I'm mostly on vacation these days.

There is no need to apologize. It's not like I don't have anything to do anymore :P

Enjoy your vacation (if you are still off :)).

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/FunctionBuilder.scala:
```diff
@@ -0,0 +1,392 @@
+    resultTypes = tpes
+
+  def setResultType(tpe: Type): Unit =
+    setResultTypes(tpe :: Nil)
```
They're not mutually exclusive. I added a longer comment on `setFunctionType` to clarify its purpose, and also to clarify why it is not exclusive with `setResultType(s)`.

Given that they're not mutually exclusive, adding them to constructor parameters would directly lead to some combinatorial explosion. The purpose of a `Builder` object is to avoid this kind of issue by admitting some mutability.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/FunctionBuilder.scala:
```diff
@@ -0,0 +1,406 @@
+  def ++=(instrs: Iterable[Instr]): Unit =
+    this.instrs ++= instrs
+
+  def markCurrentInstructionIndex(): InstructionIndex =
+    new InstructionIndex(instrs.size)
+
+  def insert(index: InstructionIndex, instr: Instr): Unit =
+    instrs.insert(index.value, instr)
```
We do create multiple marks that are live at the same time. What matters is that they're created and used in a stack order. For example "create 1, create 2, use 2, use 1" is valid, but "create 1, create2, use1, use 2" is not.

gzm0 replied:
It looks like this is only ever used to insert "conditional blocks". Do you think we could amend the API to something like (maybe with additional params for the block type):

```
def maybeBlock(body: => Unit)(cond: => Boolean)(tail: => Unit): Unit
```

IIUC this would:
- Cover existing usage (without really impeding control flow at the usage site).
- Enforce that marks are used in the right order.

sjrd replied:
That doesn't work as nicely as that. If you look at the call site in `genLabeled`, there are two things that make this problematic:
* The `label` of the optional `Block` can only be obtained if the `cond` succeeds (because of the call to `requireCrossInfo()`).
* There are conditional instructions even *after* the optional block's `end`.

So the signature would have to be
```scala
def maybeBlock(body: => Unit)(cond: => Boolean)(blockLabel: Option[LabelID])(
    tailInsideBlock: =>Unit)(tailAfterBlock: => Unit): Unit
```
which becomes super specialized to its call site, at that point.

gzm0 replied:
Fair enough. I guess we'll have to live with it :)

sjrd reviewed (commented):
Addressing some more comments.

Main todos:
* Moving from `Emitter` to `ClassEmitter` stuff in the `start` function (not yet convinced that would be better)
* Extracting `StringPool`
* Comments in `TextWriter`

sjrd reviewed (commented):
Phew, I think I have addressed everything so far.

sjrd reviewed (commented):
Oops, I had forgotten a bunch of comments in `TextWriter`.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Preprocessor.scala:
```diff
@@ -0,0 +1,458 @@
+     * Two spine types can share the same bucket only if they do not have any
+     * common join type descendants.
+     *
+     * Visit spine types in reverse topological order because (from leaves to
+     * root) when assigning a spine type to a bucket, the algorithm already has
```
```suggestion
     * Visit spine types in reverse topological order (from leaves to root)
     * because when assigning a spine type to a bucket, the algorithm already has
```

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Preprocessor.scala:
```diff
@@ -0,0 +1,458 @@
+      val ifaces = getAllInterfaces(clazz)
+
+      if (ifaces.nonEmpty) {
+        val joins = joinsOf.getOrElse(clazz.name.name, new mutable.HashSet())
+
+        if (joins.nonEmpty) {
```
```suggestion
        joinsOf.get(clazz.name.name) match {
          case Some(joins) =>
            assert(joins.nonEmpty)
            // ...
          case None =>
```

Feels odd to create a set merely to then check it is empty.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Preprocessor.scala:
```diff
@@ -0,0 +1,458 @@
+    val spines = new mutable.HashSet[ClassName]()
+
+    // Phase 1: Assign buckets to spine types
+    for (clazz <- classes.reverseIterator) {
+      val ifaces = getAllInterfaces(clazz)
+
+      if (ifaces.nonEmpty) {
```
```suggestion
    for {
      clazz <- classes.reverseIterator
      val ifaces = getAllInterfaces(clazz)
      if ifaces.nonEmpty
    } {
```

?

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Preprocessor.scala:
```diff
@@ -0,0 +1,458 @@
+    }
+
+    // Phase 2: Assign buckets to non-spine types (plain and join types)
+    for (clazz <- classes) {
+      val ifaces = getAllInterfaces(clazz)
+
+      if (ifaces.nonEmpty && !spines.contains(clazz.name.name)) {
```
```suggestion
    for {
      clazz <- classes
      val ifaces = getAllInterfaces(clazz)
      if ifaces.nonEmpty && !spines.contains(clazz.name.name)
    } {
```

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Preprocessor.scala:
```diff
@@ -0,0 +1,458 @@
+
+  private final class Bucket {
+    /** The list of interfaces that belong to this bucket. */
+    var interfaces: List[ClassName] = Nil
```
This is never read except to find the bucket per class. Instead of keeping a list here, and then re-shaping again in `preprocess`, just keep a `mutable.Map[ClassName, Int]` in `computeItableBuckets` which you return at the end?

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Preprocessor.scala:
```diff
@@ -0,0 +1,458 @@
+      val ifaces = getAllInterfaces(clazz)
+
+      if (ifaces.nonEmpty && !spines.contains(clazz.name.name)) {
+        val used = usedOf.getOrElse(clazz.name.name, new mutable.HashSet())
```
Shouldn't this be `getOrElseUpdate`? IIUC this is a bug?

Also: Shouldn't this be just an assign? We are not assigning `usedOf` for spines above (IIUC we can't), so we should hit each class exactly once?

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Preprocessor.scala:
```diff
@@ -0,0 +1,458 @@
+        val used = usedOf.getOrElse(clazz.name.name, new mutable.HashSet())
+        for {
+          iface <- ifaces
+          parentUsed <- usedOf.get(iface)
```
Besides `usedOf` never being mutated, IIUC this will not contain spine types? Is that OK? If yes, comment?

Also, doing this for all and not just direct ancestors feels odd. Direct ancestors already should have transitively included buckets of their ancestors, no?

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Preprocessor.scala:
```diff
@@ -0,0 +1,458 @@
+    }
+  }
+
+  /** Group interface types and types that implement any interfaces into buckets,
```
@tanishiking thank you for writing this explanation. Otherwise this would have been much harder to understand.

gzm0 reviewed (changes requested):
Review of everything except BinaryWriter and the points I have explicitly left comments (IIRC, primarily the dragons).b

sjrd reviewed (commented):
Addressed some comments. I need to look more deeply into the possible bug in `computeItableBuckets`.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/WasmContext.scala:
```diff
@@ -0,0 +1,317 @@
+    def registerDynamicCall(methodName: MethodName): Unit =
+      methodsCalledDynamically += methodName
+
+    def buildMethodTable(): Unit = {
```
So as I mentioned in another comment, not this one as long as `isEffectivelyFinal` is required.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/WasmContext.scala:
```diff
@@ -0,0 +1,317 @@
+    private var _specialInstanceTypes: Int = 0
+
+    def addSpecialInstanceType(jsValueType: Int): Unit =
+      _specialInstanceTypes |= (1 << jsValueType)
```
This one is definitely possible but a bit annoying. I'll try again later.

sjrd commented:
Well ... it turns out we didn't need Phase 2 of the bucket assignment algorithm at all! I added a few commits that should address the latest comments on that algorithm.

@tanishiking Could you double-check what I did there?

tanishiking commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Preprocessor.scala:
```diff
@@ -401,15 +412,11 @@ object Preprocessor {
-     * compute the set of buckets already used by its ancestors. Assign the
-     * type to any available bucket not in this set. If no available bucket
-     * exists, create a new one.
+     * Since we only need to assign itable indices to interfaces, and we
+     * observed that interfaces are all spine types, we can entirely skip this
+     * phase of the paper's algorithm.
```
💯

tanishiking reviewed (commented):
Thanks, I left one question, and otherwise the change looks reasonable :+1:

> It's interesting to note that there are barely more buckets for the full test suite than for the helloworld. We basically get 40 buckets because of the huge mess that is the collections hierarchy, but not much more beyond that.

Yeah, that happens when a class has a bunch of superclasses/interfaces, as all those superclasses share the same subclass and therefore can't be assigned to the same bucket. This situation applies to the hierarchy in the Scala collection libraries :/

I believe most applications have that number of buckets, but not many go beyond that, as long as the applications define a more deeply nested data hierarchy than the Scala collections.

<img width="457" alt="shapes at 24-07-31 13 34 54" src="https://github.com/user-attachments/assets/a261107d-4148-4e87-8122-13215aa8d8ae">

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Preprocessor.scala:
```diff
@@ -382,6 +383,16 @@ object Preprocessor {
      *
+     * Now observe that:
+     *
+     * - we only work with types that have instances,
+     * - the only way an *interface* `I` can have instances is if there is a
+     *   *class* with instances that implements it,
+     * - there must exist such a class `C` that is a join type: one that
+     *   extends another *class* and also at least one interface that has `I`
+     *   in its ancestors (note that `jl.Object` implements no interface),
```
> So, am I correct that the "another class" is the class that is descendant of `j.l.Object`?

Yes, a non-interface `class` must extend another non-interface `class`, and the chain continues until we get to `java.lang.Object`.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/BinaryWriter.scala:
```diff
@@ -0,0 +1,660 @@
+      while (i != len) {
+        byte(utf8(i))
+        i += 1
+      }
```
With ByteBuffers, we could improve this by adding a `writeTo` method on `UTF8String`.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/BinaryWriter.scala:
```diff
@@ -0,0 +1,660 @@
+      sourceMapWriter.endNode(buf.currentGlobalOffset)
+
+    override protected def emitSourceMapSection(buf: Buffer): Unit = {
+      writeCustomSection(buf, "sourceMappingURL") { buf =>
```
Ref this: https://github.com/WebAssembly/tool-conventions/blob/main/Debugging.md#source-maps ?

gzm0 reviewed (changes requested):
Review of BinaryWriter (mostly minor things) & follow-up changes (IIRC no additional comments).

At this point, all remaining things to be done are in unresolved conversations (some of them are TODOs for myself).

sjrd reviewed (commented):
Addressed comments in `BinaryWriter`, but not elsewhere yet.

sjrd commented:
And I addressed the remaining comments that were not resolved yet, AFAICT.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```suggestion
          // The order does not matter here because they will be "re-sorted" by emitBRTable
```

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```suggestion
   * we cannot use the stack for the `try_table` itself: each label has
```

gzm0 reviewed (changes requested):
Complete review. The dragons seem to be implemented nicely. So it is not that terrible to understand :)

sjrd reviewed (commented):
Addressed latest batch of comments.

gzm0 reviewed (approved):
LGTM after squashing (down to 4 total commits).

sjrd commented:
Thank you so much @gzm0 for this epic review!

sideeffffect commented on linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/TextWriter.scala:
> binaryen does not support inputs or multi-value outputs

It should support it now :tada: 
* https://github.com/WebAssembly/binaryen/pull/7149
